framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceImplementation.java

```diff
@@ -662,6 +662,19 @@ protected void updateAnnotationSet(
     if (rhsATM instanceof AnnotatedNullType && ignoreNullAssignments) {
       return;
     }
+
+    // If the rhsATM and the lhsATM have different kinds (which can happen e.g. when
+    // an array type is substituted for a type parameter), do not attempt to update
+    // the inferred type, because this method is written with the assumption
+    // that rhsATM and lhsATM are the same kind.
+    if (rhsATM.getKind() != lhsATM.getKind()) {
+      // The one difference in kinds situation that this method can account for is the RHS being
```

Comment:
One possibility is to forbid all cases where the kinds differ, and add a few exceptions.
The other possibility is to only forbid certain cases.  The latter is less likely to lead to difficult-to-debug problems.  It is also more likely to lead to future crashes that we must address, but at least we'll know about them rather than them being silent failures.  Therefore, the second possibility seems better to me.

Reply:
I considered the approach you suggested, and chose this approach for two reasons:
1. simplicity: there are more cases that we want to forbid than cases that we want to permit, so rejecting anything that isn't a specifically-listed exception results in simpler code, and
2. WPI's specification generally permits it to fail to make a true inference, except in the cases we know to work (and test). So, I preferred to fail to make (some) inferences rather than crashing, which prevents WPI from being used at all and therefore fails to make any inferences.

I'm open to changing to the other approach. Do either of the reasons above change your feeling that the allow-by-default approach is better than the current deny-by-default approach I've proposed?

Reply:
@mernst, do you have an opinion on this? I'm assigning this PR to you, because I'd like to discuss what the right behavior for WPI in this case actually is.